### PR TITLE
[ICDS Dashboard]: Poshan Progress Dashboard API

### PR DESCRIPTION
**Pull request by ShubhamBansal1997**
##### SUMMARY
https://app.asana.com/0/1177124110930377/1177375896402662/f

### CODE REVIEW

File: custom/icds_reports/sqldata/exports/poshan_progress_report.py [hunk 1]
```diff
@@ -67,6 +67,33 @@
 }
 
 
+def _generate_quarter_months(quarter, year):
```
Comment (Rohit25negi) [inline custom/icds_reports/sqldata/exports/poshan_progress_report.py:70]: why is this changed?

File: custom/icds_reports/reports/poshan_progress_dashboard_data.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        # for quarter we need to average summation
+        if data_format == 'quarter':
+            for k, v in aggregated_row.items():
+                if k not in ['state_name', 'district_name', 'district_id', 'state_id']:
+                    aggregated_row[k] = _handle_average(v)
+        aggregated_row = calculate_percentage_single_row(deepcopy(aggregated_row))
+        aggregated_row = prepare_structure_aggregated_row(deepcopy(aggregated_row), count, aggregation_level)
+        return aggregated_row
+
+    def prepare_structure_aggregated_row(row, count, aggregation_level):
```
Comment (Rohit25negi) [inline custom/icds_reports/reports/poshan_progress_dashboard_data.py:74]: good to pull all these fuctions out of the main function and make then protected to this file instead of making themnested functions

File: custom/icds_reports/reports/poshan_progress_dashboard_data.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
+from copy import deepcopy
+from datetime import date
+
+from custom.icds_reports.cache import icds_quickcache
+from custom.icds_reports.models.views import PoshanProgressReportView
+from custom.icds_reports.sqldata.exports.poshan_progress_report import (
```
Comment (Rohit25negi) [inline custom/icds_reports/reports/poshan_progress_dashboard_data.py:6]: if you want to use some code that is shared between reports and this. Good to move that code to a central util file instead of doing this

**Review by Rohit25negi — CHANGES_REQUESTED** [review]
could you please restructure the code as mentioned in comments to ease out the review.
Also, these summations done in python can it be done while fetching the records itself through ORM

File: custom/icds_reports/reports/poshan_progress_dashboard_data.py (hunk 3 above)
Comment (ShubhamBansal1997) [reply custom/icds_reports/reports/poshan_progress_dashboard_data.py:8]: updated

File: custom/icds_reports/reports/poshan_progress_dashboard_data.py (hunk 2 above)
Comment (ShubhamBansal1997) [reply custom/icds_reports/reports/poshan_progress_dashboard_data.py]: updated

File: custom/icds_reports/sqldata/exports/poshan_progress_report.py (hunk 1 above)
Comment (ShubhamBansal1997) [reply custom/icds_reports/sqldata/exports/poshan_progress_report.py]: to make it useful for both the files

File: custom/icds_reports/reports/poshan_progress_dashboard_data.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    }
+    return data
+
+
+def calculate_comparitive_rows(data, aggregation_level, data_format, unique_id):
+    # for quarter we need to average summation
+    quarter_compartivie_dict = {}
+    if data_format == 'quarter':
+        for i in range(0, len(data)):
+            if data[i][unique_id] in quarter_compartivie_dict.keys():
```
[the remark this answers is not in the document]
Comment (Rohit25negi) [reply custom/icds_reports/reports/poshan_progress_dashboard_data.py:117]: should the condition be `not in`?

**Review by Rohit25negi — CHANGES_REQUESTED** [review]
Looks much better now. 
left small comments.

File: custom/icds_reports/reports/poshan_progress_dashboard_data.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ShubhamBansal1997) [reply custom/icds_reports/reports/poshan_progress_dashboard_data.py]: updated

**Review by Rohit25negi — APPROVED** [review]
test case failure looks related

File: custom/icds_reports/reports/poshan_progress_dashboard_data.py [hunk 5]
```diff
@@ -175,7 +186,7 @@ def get_poshan_progress_dashboard_data(domain, year, month, quarter, data_format
 
     if step == 'aggregated':
         data = queryset.aggregate(
-            count=Count('state_name'),
+            count=Count('state_name', distinct=True),
```
Comment (Rohit25negi) [inline custom/icds_reports/reports/poshan_progress_dashboard_data.py:189]: could you please tell why this is hardcoded to state_name?

**Review by Rohit25negi — COMMENTED** [review]
left one question, otherwies looks good

File: custom/icds_reports/reports/poshan_progress_dashboard_data.py (hunk 5 above)
Comment (ShubhamBansal1997) [reply custom/icds_reports/reports/poshan_progress_dashboard_data.py:189]: oh because state_name are needed, when it's national level

Comment (Rohit25negi) [reply custom/icds_reports/reports/poshan_progress_dashboard_data.py:189]: and when its state level, dont we want district names?

**Review by Rohit25negi — APPROVED** [review]

**Comment by ShubhamBansal1997** [comment]
https://travis-ci.org/github/dimagi/commcare-hq/builds/700115911 Build Passed
